src/libp2p/peers.rs

```diff
@@ -370,21 +388,36 @@ where
                 }
 
                 collection::Event::StartShutdown { id, reason } => {
-                    // TODO: this is O(n)
-                    for (_, item) in &mut self.desired_out_notifications {
-                        if let Some((_, connection_id, _)) = item.as_ref() {
-                            if *connection_id == id {
-                                *item = None;
-                            }
-                        }
-                    }
-
                     let connection_state = self.inner.connection_state(id);
                     debug_assert!(connection_state.shutting_down);
 
                     let peer = if let Some(peer_index) = self.inner[id].peer_index {
                         let peer_id = self.peers[peer_index].peer_id.clone();
 
+                        // We might have to insert the peer back in `unfulfilled_desired_peers` if
+                        // it is desired.
+                        if (self.peers[peer_index].desired
```

Comment:
do you want to extract this into a function to avoid duplication?

Reply:
So there are 4 blocks that are similar, but only two of the four are actually exactly the same, and for these 2 it's not just this block that is repeated but way more code. I want to tackle this separately afterwards.